### DISCUSSION TITLE
Atom creator for reactive values

### DISCUSSION THREAD

**Discussion by venkatd** [pmndrs/jotai#1844] [Ideas]
Integrating jotai with the outside, messy event-based stateful requires you to be an intermediate/expert in jotai.

For example:
- An atom to model `window.location` as seen in the jotai-location package
- The current time
- The browser window size
- Online/offline state

All of these have some properties in common:
- They are reactive (with a handler or callback, they will notify us when the value has changed)
- They need disposal (unsubscribing to events, clearInterval, etc)
- In some cases (such as location), the atom may also be read/write.
- A reactive value from a client-side database such as tinybase (get, set, and updates when the db value updates)

Ideally we can just implement this declaratively without thinking much about low-level jotai concepts.

An example of a basic usage could look like this:

```ts
type TimeAction =
  | { type: 'log_time'; format: string }
  | { type: 'rewind_time'; to: Date }

const currentTimeAtom = statefulAtom((onUpdate) => {
  var currentTime = new Date()

  const id = setInterval(() => {
    currentTime = new Date()
    onUpdate()
  }, 1000)

  return {
    get: () => currentTime,
    set: (action: TimeAction) => {
      // we could potentially dispatch actions with side-effects here
    },
    unmount: () => clearInterval(id)
  }
})
```

The code above is just a rough sketch. I'm not feeling strongly about the syntax or the name, but sharing to start a brainstorm for the best solution!

**Comment by dai-shi** [comment]
Interesting idea. It's not really like an atom model I had in mind, but it will opens up some new use cases.
Actually, we have a nice undocumented feature in v2 for this.

Side effects are probably fine as long as the update is not in sync, we can delay by default.
Bigger issue is atom value must be an immutable object. So, I would freeze object by default.

Here's the rough idea:
```ts
const deepFreeze = (obj: any) => {
  if (typeof obj !== 'object' || obj === null) return
  Object.freeze(obj)
  const propNames = Object.getOwnPropertyNames(obj)
  for (const name of propNames) {
    const value = obj[name]
    deepFreeze(value)
  }
  return obj
}

export function statefulAtom<T>(
  create: (onUpdate: (nextValue: T) => void) => T,
  options?: {
    dangerously_disableFreezing: boolean;
    dangerously_disableDelaying: boolean;
  }
): Atom<T> {
  const baseAtom = atom(null)
  const modelAtom((get, { setSelf }) => {
    const base = get(baseAtom)
    if (base) {
      return baseAtom
    }
    const initialValue = create((nextValue) => {
      if (!options?.dangerously_disableFreezing) {
        deepFreeze(nextValue)
      }
      if (!options?.dangerously_disableDelaying) {
        Promise.resolve().then(() => setSelf(nextValue))
      } else {
        setSelf(nextValue)
      }      
    })
    if (!options?.dangerously_disableFreezing) {
      deepFreeze(initialValue)
    }
    return initialValue
  }, (get, set, nextValue) => {
    set(baseAtom, nextValue)
  )
  return atom((get) => get(modelAtom))
}
```

The usage is a bit tricky because of freezing.

```js
const currentTimeAtom = statefulAtom((onUpdate) => {
  let currentTime = new Date()

  const get = () => currentTime
  const set = (action) => ...
  const unmount = () => ...

  const id = setInterval(() => {
    currentTime = new Date()
    onUpdate({ get, set, unmount })
  }, 1000)

  return { get, set, unmount }
})
```

We could use `Object.assign`.

A caveat: as long as `get` function is used, it's reading a ref value, so using it in React isn't concurrent safe. It's a trade off. I think it works mostly fine.
Another caveat: no functions should be called during render. That rule should be followed anyways, so "delaying" might not be needed.

**Reply by venkatd** [reply]
Very cool!

> Side effects are probably fine as long as the update is not in sync

Do you mean as long as the update isn't triggered synchronously?

> Bigger issue is atom value must be an immutable object.
Could you clarify this one? Do you mean the `{get, set, unmount}` references must be immutable?

> it's reading a ref value, so using it in React isn't concurrent safe
Could you clarify this one?

What is the reason for `onUpdate` passing along `get`, `set`, and `unmount`? Don't we only need to notify jotai that the value has changed?

**Reply by dai-shi** [reply]
> Do you mean as long as the update isn't triggered synchronously?

Yes.

> What is the reason for `onUpdate` passing along `get`, `set`, and `unmount`? Don't we only need to notify jotai that the value has changed?

Because the atom value must be immutable.

Hm, that gives me another idea.

```js
function atomWithMethods(initialValue, createMethods) {
  const baseAtom = atom(initialValue);
  const methodsAtom = atom(
    (get, { setSelf }) => createMethods(setSelf),
    (get, set, nextValue) => set(baseAtom, nextValue)
  );
  return atom((get) => {
    const value = get(valueAtom);
    const methods = get(methodsAtom);
    return { ...methods, value };
  });
}
```

Usage:

```js
const currentTimeAtom = atomWithMethods(0, (setValue) => {
  const id = setInterval(() => {
    const currentTime = new Date()
    setValue(currentTime)
  }, 1000)

  return {
    set: (action: TimeAction) => {
      // we could potentially dispatch actions with side-effects here
    },
    unmount: () => clearInterval(id)
  }
})
```

It's different from stateful model outside jotai, though.

**Comment by venkatd** [comment]
I've managed to get a proof of concept working.

**Here is the usage:**
```ts
import { atomWithObservable } from '@util/jotai'
import { useAtom } from 'jotai'

type SWAction = 'play' | 'pause' | 'reset'

const stopWatchAtom = atomWithObservable<number, [SWAction], void>(() => {
  const empty = () => {}

  let updated = empty

  let elapsed: number = 0

  let timer: NodeJS.Timer | undefined

  return {
    get: () => elapsed,
    addListener: (l) => {
      updated = l
      return () => {
        updated = empty
      }
    },
    set: (action: SWAction) => {
      switch (action) {
        case 'play':
          if (timer) clearInterval(timer)
          timer = setInterval(() => {
            elapsed += 1
            console.log('tick')
            updated()
          }, 1000)
          break
        case 'reset':
          elapsed = 0
          updated()
          break
        case 'pause':
          elapsed = 0
          if (timer) clearInterval(timer)
          updated()
          break
      }
    },
    unmount: () => {
      if (timer) clearInterval(timer)
    }
  }
})

export function SWSandbox() {
  const [elapsed, dispatch] = useAtom(stopWatchAtom)
  return (
    <div>
      <button onClick={() => dispatch('play')}>play</button>
      <button onClick={() => dispatch('pause')}>pause</button>
      <button onClick={() => dispatch('reset')}>reset</button>
      <br />
      <b>{elapsed}</b>
    </div>
  )
}
```

**Implementation below**
```ts
import { atom, Atom, WritableAtom } from 'jotai/vanilla'

type Listener = () => void
type Unsubscribe = () => void
type Dispose = () => void

type Read<Value> = Atom<Value>['read']

export interface Observable<Value, Args extends unknown[], Result> {
  get: () => Value
  addListener(listener: Listener): Unsubscribe
  unmount(): void
  set: (...args: Args) => Result
}

type Action = 'subscribe'

export function atomWithObservable<
  Value,
  Args extends unknown[] = void[],
  Result = void
>(
  read: Read<Observable<Value, Args, Result>>
): WritableAtom<Value, Args, Result> {
  const anAtom = atom(read)
  return toObservableAtom(anAtom)
}

export function toObservableAtom<Value, Args extends unknown[], Result>(
  baseAtom: Atom<Observable<Value, Args, Result>>
): WritableAtom<Value, Args, Result> {
  const refreshCounter = atom(0)

  let dispose: Dispose | undefined

  let lastObservable: Observable<Value, Args, Result> | undefined
  let lastDispatch: (action: Action) => void | undefined

  const innerObservableAtom = atom<
    Observable<Value, Args, Result>,
    [Action],
    void
  >(
    (get) => {
      const observable = get(baseAtom)
      if (observable == lastObservable) return observable

      // Workaround for ensuring we resubscribe when the base atom changes
      // onMount is only called once upon creation, so this workaround
      // is necessary
      lastObservable = observable
      // To prevent an update during render (which causes a React error)
      // we wait until the next frame to subscribe and trigger a refresh
      requestFrame(() => lastDispatch?.('subscribe'))

      return observable
    },
    (get, set) => {
      const observable = get(innerObservableAtom)

      const subscribe = () => {
        set(refreshCounter, (x) => x + 1)
      }

      dispose?.()

      const unsubscribe = observable.addListener(subscribe)
      dispose = () => {
        unsubscribe?.()
        observable.unmount()
      }
    }
  )
  innerObservableAtom.debugPrivate = true
  innerObservableAtom.onMount = (dispatch) => {
    lastDispatch = dispatch

    dispatch('subscribe')

    return dispose
  }

  const stateAtom = atom<Value, Args, Result>(
    (get) => {
      get(refreshCounter) // force a cache update
      return get(innerObservableAtom).get()
    },
    (get, set, ...args: Args) => {
      const observable = get(innerObservableAtom)
      return observable.set(...args)
    }
  )
  stateAtom.debugPrivate = true

  return stateAtom
}

function requestFrame(cb: () => void) {
  if (typeof window === 'undefined') return cb() // prevent crash on ssr
  requestAnimationFrame(cb)
}
```

Some points:
- I picked the name `atomWithObservable` but since that conflicts with the built in util, it's probably better to come up with a different name.
- There are some hacks which I needed to get it working, but it seems to work as expected in some demos
- I'm not crazy about how the `addListener` implementation is done. Perhaps an `onChange` could be a 2nd argument to the atom so the user doesn't have to think about managing listeners.
- Part of the goal is to make it straightforward to define non-pure atoms. Actually, I think it may be better practice to not have large atom definitions. For example, with the example above, we could define a `Stopwatch` class which handles most of this logic and expose it to jotai through `atomWithObservable`

**Reply by dai-shi** [reply]
> * I picked the name `atomWithObservable` but since that conflicts with the built in util, it's probably better to come up with a different name.

Yeah, it's too confusing for me.

> * There are some hacks which I needed to get it working, but it seems to work as expected in some demos

What should be fixed is variables outside atoms, namely `dispose`, `lastObservable` and `lastDispatch`.
They don't work well (they conflict) with multi-store use case.

> * Part of the goal is to make it straightforward to define non-pure atoms.

Yeah, this is for non-pure atoms. (atomWithRefresh is similar).
We don't add it  in jotai utils, but as long as you know it, it should be fine.

**Reply by venkatd** [reply]
> Yeah, it's too confusing for me.

Maybe `statefulAtom` is better

> They don't work well (they conflict) with multi-store use case.

What might a good solution for this be? Are there any good examples out there for handling it?

> We don't add it in jotai utils, but as long as you know it, it should be fine.

Yes, I understand the desire to keep the core of jotai small!

**Reply by dai-shi** [reply]
> What might a good solution for this be? Are there any good examples out there for handling it?

Basically everything has to be atoms.
Check out some existing utils in `jotai/utils` and maybe `jotaijs/*`.

Technically the `read` function has to be pure, so we should expect the function to be invoked multiple times. It's limited like React render and/or useEffect.
There would be may patterns for hacks..
